### ISSUE TITLE
Bug: iOS VoiceOver announces stale content in headings

### ISSUE THREAD

**Issue by JeffreyATW** [facebook/react#21212]
React version: 17.0.2

## Steps To Reproduce

1. Go to https://jeffreyatw.com/static/voiceover-react using iOS, enable VoiceOver
2. Swipe to first "Not clicked" h2, observe VoiceOver reads content correctly
3. Click the first "click me" button
4. Move VoiceOver's focus back to first h2, which is now "Clicked"
5. VoiceOver **incorrectly** reads it as "Not clicked, heading level 2"
6. Swipe to the second "Not clicked" h2, observe VoiceOver reads content correctly
7. Click the second "click me" button
8. Move VoiceOver's focus back to second h2, which is now "Clicked"
9. VoiceOver correctly reads it as "Clicked, heading level 2"

Link to code example: https://jeffreyatw.com/static/voiceover-react

## The current behavior
VoiceOver reads the first h2 as "Not clicked, heading level 2". See attached video for more details:
https://user-images.githubusercontent.com/266170/114103859-b4b51680-987e-11eb-82b9-0d785442c3ae.mp4

## The expected behavior
VoiceOver should read the first h2 as "Clicked, heading level 2"

**Comment by eps1lon** [comment]
Thanks for the report.

In order to understand the issue, it would help if you could post the code that produced https://jeffreyatw.com/static/voiceover-react/. It's otherwise ambigious what you mean by "with react" and "without React".

Though I suspect that this should be reported to VoiceOver as well.

For performance reasons React doesn't set `textContent` but uses text nodes. Your "without react" example is probably using `textContent` and VoiceOver is only listening to changes to `textContent`. But I can't know for certain without seeing the code.

**Comment by JeffreyATW** [comment]
Sure. The code is unminified on the page I'm hosting, but here it is for easier perusal:

```html
<script
  src="https://unpkg.com/react@17/umd/react.development.js"
  crossorigin
></script>
<script
  src="https://unpkg.com/react-dom@17/umd/react-dom.development.js"
  crossorigin
></script>

<!-- Load our React component. -->
<h1>With React</h1>
<div id="root"></div>

<h1>Without React</h1>
<h2 id="heading">Not clicked</h2>
<button id="button">Click me</button>

<script type="text/javascript">
  document.getElementById("button").addEventListener("click", () => {
    document.getElementById("heading").textContent = "Clicked";
  });
</script>

<script src="script.js"></script>
```

```js
"use strict";

const e = React.createElement;

const Foo = () => {
  const [clicked, setClicked] = React.useState(false);

  return e(React.Fragment, {}, [
    e("h2", { key: "1" }, clicked ? "Clicked" : "Not clicked"),
    e("button", { key: "2", onClick: () => setClicked(true) }, "Click me"),
  ]);
};

const domContainer = document.querySelector("#root");
ReactDOM.render(e(Foo), domContainer);
```

So yes, I am using `textContent` with vanilla JS. But the point still stands - VoiceOver is reading stale heading content and I don't know how to force it to read the new one.

I'm not interested in using `aria-live` regions since it doesn't solve the issue of the heading itself being read wrong when swiping over to it.

**Comment by JeffreyATW** [comment]
I investigated further and it does seem that [`setTextContent`](https://github.com/facebook/react/blob/9198a5cec0936a21a5ba194a22fcbac03eba5d1d/packages/react-dom/src/client/setTextContent.js) changing the `nodeValue` instead of `textContent` is what causes this issue.

So unless this is expected behavior on VoiceOver's part, this is not a React bug. Feel free to close this if you agree; I can file a bug with bugs.webkit.org.

**Comment by JeffreyATW** [comment]
Looks like it's already an open bug here. https://bugs.webkit.org/show_bug.cgi?id=190437

**Comment by eps1lon** [comment]
> I investigated further and it does seem that setTextContent changing the nodeValue instead of textContent is what causes this issue.

Thanks for confirming https://github.com/facebook/react/issues/21212#issuecomment-816279500

> Looks like it's already an open bug here. bugs.webkit.org/show_bug.cgi?id=190437

Appreciate the reference. Might be interesting to know how much perf gain `textContent` over changing `nodeValue` has. Though in the end VO should really fix this since React is just using standard DOM API. Maybe knowing that React sets `nodeValue` VO will bump the priority.

Leaving this open for visibility unless React core has other plans.

**Comment by marcysutton** [comment]
I'm seeing (potentially) the same VoiceOver issue on custom Combobox openers when rendered with React. Here is a minimal reproduction: https://codepen.io/marcysutton/pen/azoEMpO?editors=0110
